core/src/replay_stage.rs

```diff
@@ -1780,6 +1790,16 @@ impl ReplayStage {
         did_complete_bank
     }
 
+    fn persist_cost_table(blockstore: &Blockstore, cost_model: &RwLock<CostModel>) {
+        let cost_model_read = cost_model.read().unwrap();
+        let cost_table = cost_model_read.get_instruction_cost_table();
+        for (key, cost) in cost_table.iter() {
+            blockstore
+                .write_program_cost(key, cost)
+                .expect("persist program costs to blockstore");
+        }
+    }
+
```

Comment:
In general, I think it's better that noncritical updates that are potentially expensive (grabbing read/write locks, updating database) are offloaded from the replay thread. 

From my understanding, the cost model mostly affects BankingStage, and ReplayStage is just feeding the model datapoints. From this perspective, it seems both `update_cost_model()` and this call to `persist_cost_table()` could be offloaded to a separate thread where updates are pushed to that thread via a channel. Is this feasible?

Reply:
This is great suggestion. `replay_stage` currently has one thread that does replay, adding a `services` thread to perform non-critical helps now and future. I'll give it a try.

Reply:
Add service thread runs `service_loop`, main thread sends `execute_timings` to it via a channel. 
`service_loop` updates cost_model; if cost table is changed, it'd sync up what's in memory with what's blockstore by deleting old programs and update/insert new ones.